### ISSUE TITLE
Modification of Values in model in EventCreationPolicy.InsertOnStartReplaceOnEnd and InsertOnStartInsertOnEnd

### ISSUE THREAD

**Issue by dammitjanet** [thepirat000/Audit.NET#73]
I've got a custom action where I need to blank out WebAPI model values when saving to the DB (passwords and large values like byte[] arrays)

As I am in InsertOnStartReplaceOnEnd creation policy, the custom action running on Action.OnEventSaving executes twice, once for the insert and once for the update (replace).

I can determine which state by looking at the EventId (null when inserting, non-null when updating), I can then blank out the password values when replacing by only updating the model values when the EventId is populated.

If however an Event never gets replaced, I have passwords in plain text in the Audit.log.

Also when the CreationPolicy is InsertOnStartInsertOnEnd I can't determine which insert (1st or 2nd) I am looking at as both EventId's are null.

For me there are two issues:

1. Values in the 'scope' appear to be 'By Reference' and are not cloned (I will assume speed plays a part here) so that any change to those values when a record is first inserted under either InsertOnStartReplaceOnEnd and InsertOnStartInsertOnEnd policies, breaks the functionality of the events being audited. 

There may be a place for some method of either:
 - decorating specific models or fields/properties on models with an Audit.net specific annotation to not log
 - decorating specific methods/events to force a clone of the objects so that altering the values that are logged does not affect the underlying method/parameters

2. Being unable to differentiate where in the event logging lifecycle the event actually is, means that there might be a place for further ActionTypes based on different content creation policies:
e.g.
 - OnEventReplacing
 - OnFirstInsert
 - OnSecondInsert 

... or there might be a place for an additional enumeration to describe the additional state within the scope:
e.g.
- AuditState.InsertFirst
- AuditState.InsertSecond
- AuditState.Replace

**Comment by thepirat000** [comment]
Hi

For 1: currently, only the custom fields can be optionally pre-serialized

https://github.com/thepirat000/Audit.NET/blob/9b927a8c1885c0f7bdd4ae4bc28bfb51aa9ddeec/src/Audit.NET/AuditScope.cs#L177

which values of the scope/event are you expecting to be cloned?

For 2: I can add a `SaveMode` property to the `AuditScope`. Maybe with values like: InsertAtStart, InsertAtEnd, ReplaceAtEnd and Manual. Will that be useful for you?

**Comment by thepirat000** [comment]
Just added the SaveMode property to the AuditScope object, this is starting from version 8.7.0

**Comment by dammitjanet** [comment]
Thanks I'll check out the SaveMode shortly.

For the serialisation this is my Audit.WebAPI config for reference

`			Audit.Core.Configuration
				.Setup()
				.UseSqlServer(config => ..[redacted]..
						)
				.WithCreationPolicy(EventCreationPolicy.InsertOnStartReplaceOnEnd);

			// custom action to blank out passwords and large image files
			Audit.Core.Configuration.AddCustomAction(ActionType.OnEventSaving, scope =>
			{
				if (scope.SaveMode == SaveMode.ReplaceOnEnd)
				{
					var apiEvent = scope.Event.GetWebApiAuditAction();
					if (apiEvent.ActionParameters.ContainsKey("model"))
					{
						var modelType = apiEvent.ActionParameters["model"].GetType();
						if (modelType == typeof(ChangePasswordViewModel))
						{
							var model = apiEvent.ActionParameters["model"] as ChangePasswordViewModel;
							model.CurrentPassword = "*";
							model.NewPassword = "*";
							model.ConfirmNewPassword = "*";
						}
						else if (modelType == typeof(SetPasswordAdminViewModel))
						{
							var model = apiEvent.ActionParameters["model"] as SetPasswordAdminViewModel;
							model.NewPassword = "*";
							model.ConfirmNewPassword = "*";
						}
						else if (modelType == typeof(SetPasswordViewModel))
						{
							var model = apiEvent.ActionParameters["model"] as SetPasswordViewModel;
							model.NewPassword = "*";
							model.ConfirmNewPassword = "*";
						}
					};

					if (apiEvent.ActionParameters.ContainsKey("image") && apiEvent.ActionParameters["image"] != null)
					{
						var imageLength = (apiEvent.ActionParameters["image"] as byte[]).Length;
						apiEvent.ActionParameters["image"] = null;
						scope.Event.CustomFields["imageSize"] = imageLength;
					};
				}
			});`

So for each of my models above, and the ActionParaemters of name 'image', I'm receiving a content that I do not want to even write to the DB for the first SaveState (Insert) but using the new code you created for me yesterday, I can at least overwrite the content on the SaveMode.ReplaceOnEnd phase.

Considering some of the byte arrays that contain the images can be big (20MB) region, this isn't ideal as it has to be written to the DB at least once, and then removed.

Ideally, I'd like to decorate the property in the (password) Model with some attribute that tells Audit.Net to redact that property when writing to the audit log in any of the phase, but does not change that property when the model is passed to the underlying method.  My example would replace any instance of the model property whether insert or replace with the value inside the Attribute (including the allowance of null values)

`	public class ChangePasswordViewModel
	{
		[Required]
		[DataType(DataType.Password)]
		[Display(Name = "Password")]
		[RegularExpression(@"(^(?=.*\d)(?=.*[a-z])(?=.*[A-Z])[\W\w]{8,}$)", ErrorMessage = ValidationErrors.PasswordError)]
		[AuditRedact("*")]
		public string NewPassword { get; set; }
...
       }`

The other alternate that I could think of, would be to decorate the method that needs to be serialised differently with another property such as in the below, the default for this would be false (i.e. the current state of affairs) but where decorated, the defined objects would be serialised before the insert so changes to the 'model' would not be persisted through to the Actions.

`	[AuditApi(EventTypeName = "{controller}/{action} ({verb})", SerialiseModeI = true, SerialiseActionParameters = true, SerialiseFormVariables = true)]
	public class PasswordApiController : BaseApiController
{
...
}`

Hope this gives you a better idea of what my thoughts are.

**Comment by thepirat000** [comment]
Oh, I see. I will take a closer look and let you know.

In the meanwtime, you could try cloning the entire event on your customaction, like:
```c#
scope.Event = JsonConvert.Deserialize(JsonConvert.Serialize(scope.Event))
```

**Comment by thepirat000** [comment]
Just added a boolean property `SerializeActionParameters` to the attribute for MVC and WebApi, to indicate whether the action arguments should be pre-serialized to the audit event. This is included starting on version 9.0.1.

Note that Form Variables does not need to be serialized since they are strings.

Feel free to re-open the issue if you're still having problems
